torch/utils/_pytree.py

```diff
@@ -116,6 +123,9 @@ def _register_pytree_node(
             "Please use to_dumpable_context and from_dumpable_context instead."
         )
 
+    if typ in SUPPORTED_NODES:
+        raise ValueError(f"{typ} is already registered as pytree node.")
```

Comment:
We do not allow registering the same type twice in C++ pytree. So we also add this for Python pytree.

Reply:
The CI is failing here:

https://github.com/pytorch/pytorch/blob/c36b31d5302d31746f3f3bd64ed8d9acd8e36155/test/export/test_export.py#L634-L698

In the test, we register the `MyDataClass` again with the extra argument `return_none_fields=True` in line 674.

https://github.com/pytorch/pytorch/blob/c36b31d5302d31746f3f3bd64ed8d9acd8e36155/test/export/test_export.py#L647

https://github.com/pytorch/pytorch/blob/c36b31d5302d31746f3f3bd64ed8d9acd8e36155/test/export/test_export.py#L674

We can bypass this test error by inheriting `MyDataClass` and registering the subclass again in the registry.

------

For argument `return_none_fields=False`, it enumerates the dataclasss attributes and adds only non-None values into the flattened result. Couldn't we always set `return_none_fields=True` and use `tree_map_only(Tensor, ...)` APIs instead?

https://github.com/pytorch/pytorch/blob/c36b31d5302d31746f3f3bd64ed8d9acd8e36155/torch/_export/utils.py#L39-L50

Any thoughts about this?

cc @zou3519
cc @ydwu4

Reply:
> Couldn't we always set return_none_fields=True?

There's some use cases for the return_none_fileds = False. Dynamo.export produces two outputs: the output (flattend) for the graph module  that will be exported, and the output (not flattend) of the optimized python program (think of it as a wrapper around the graph module). Export need to produce a match between them to preserve the calling convention and have correct meta data and this is done by pytree-flatten the non-flattend output. This works fine for normal pytrees like list/tuple/dict. But for dataclasses, dynamo will prune the none output of the exported graph module causing a inconsistency between the flattened result of original dataclass and graph module's output. The return_none_fields = False is added because of that. It may be desirable to refactor the matching logic so that it doesn't depends on pytree but we're not sure if it's worth the effort.